source/byte_buf.c

```diff
@@ -380,6 +384,118 @@ int aws_byte_buf_append(struct aws_byte_buf *to, const struct aws_byte_cursor *f
     return AWS_OP_SUCCESS;
 }
 
+int aws_byte_buf_append_with_lookup(
+    struct aws_byte_buf *to,
+    const struct aws_byte_cursor *from,
+    const uint8_t *lookup_table) {
+    assert(from->ptr);
+    assert(to->buffer);
+
+    if (to->capacity - to->len < from->len) {
+        return aws_raise_error(AWS_ERROR_DEST_COPY_TOO_SMALL);
+    }
+
+    for (size_t i = 0; i < from->len; ++i) {
+        to->buffer[to->len + i] = lookup_table[from->ptr[i]];
+    }
+
+    to->len += from->len;
+    return AWS_OP_SUCCESS;
+}
+
+int aws_byte_buf_append_dynamic(struct aws_byte_buf *to, const struct aws_byte_cursor *from) {
+    assert(from->ptr);
+    assert(to->buffer);
+
+    if (to->capacity - to->len < from->len) {
+        /*
+         * NewCapacity = Max(OldCapacity * 2, OldCapacity + MissingCapacity)
+         */
+        size_t missing_capacity = from->len - (to->capacity - to->len);
+
+        size_t required_capacity = 0;
+        if (aws_add_size_checked(to->capacity, missing_capacity, &required_capacity)) {
+            return AWS_OP_ERR;
+        }
+
+        /*
+         * It's ok if this overflows, just clamp to max possible.
+         * In theory this lets us still grow a buffer that's larger than 1/2 size_t space
+         * at least enough to accommodate the append.
+         */
+        size_t growth_capacity = aws_add_size_saturating(to->capacity, to->capacity);
+
+        size_t new_capacity = required_capacity;
+        if (new_capacity < growth_capacity) {
+            new_capacity = growth_capacity;
+        }
+
+        /*
+         * Attempt to resize - we intentionally do not use reserve() in order to preserve
+         * the (unlikely) use case of from and to being the same buffer range.
+         */
+
+        /*
+         * Try the max, but if that fails and the required is smaller, try it in fallback
+         */
+        uint8_t *new_buffer = aws_mem_acquire(to->allocator, new_capacity);
```

Comment:
You could use `aws_mem_realloc()`. It already does a lot of this copy-from-old-to-new-buffer stuff for you.
Also, if it could theoretically be faster in the case that the existing allocation can simply grow in place

... ughh
but that wouldn't work if `from` used memory from `to`. You'd need to detect that cause and do pointer math to correct the from pointer